Implémenté LogiqueMayor, LogiqueOuvrier, l'array choixBatiment et premier exemple de Paral. Le tous dans main.go pour l'instant.

Nost22 opened the pull request.

Commit:
Implémenté LogiqueMayor, LogiqueOuvrier, l'array choixBatiment et premier exemple de Paral. Le tous dans main.go pour l'instant.

Nost22 committed:
Generate up-to-date rapport.pdf

gcleroux commented on main.go:
```diff
@@ -2,10 +2,96 @@ package main
+		if budget > choixBatiments[batimentChoisi].prixBatiment {
+			commande := choixBatiments[batimentChoisi]
+			budget = budget - commande.prixBatiment
+			OuvrierConstruit(commande)
+		}
+	}
+	return budget
+}
+
+func MayorEnd(budget int) {
+	fmt.Println("Le Mayor prend sa retraite avec un budget restant de " + strconv.Itoa(budget) + "$")
+	fmt.Println("La ville contient les bâtiments suivants:")
+	for i := 0; i < len(villeContenu); i++ {
+		fmt.Println(villeContenu[i])
+	}
+}
+
+// --- Fonctions des Ouvriers ---
+func OuvrierConstruit(commande batimentInfo) {
+	var effortTotal = commande.effortBatiment
+	fmt.Println("L'ouvrier commence la construction de : ", commande.nomBatiment)
+	for i := 0; i < effortTotal; i++ {
+		time.Sleep(time.Millisecond * 50)
+	}
+	fmt.Println("L'ouvrier a terminer la construction de : ", commande.nomBatiment)
+	villeContenu = append(villeContenu, commande.nomBatiment)
+}
+
+//--- Fonctions des Citoyens ---
+
+// --- Fonctions Globales ---
+func TrouveBatimentMoinsCher() int {
+	plusPetitPrix := math.MaxInt
+	for i := 0; i < len(choixBatiments); i++ {
+		if choixBatiments[i].prixBatiment < plusPetitPrix {
+			plusPetitPrix = choixBatiments[i].prixBatiment
+		}
+	}
+	return plusPetitPrix
 }
```
`main.go` est juste un point d'entrée.

Aucune logique devrait se trouver ici, toutes les fonctions devraient être dans `mayor.go`

Nost22 replied:
I agree, j'ai mentionné dans le nom du commit que c'est juste "pour l'instant"

gcleroux commented on main.go:
```diff
@@ -2,10 +2,96 @@ package main
+	nomBatiment    string
+	prixBatiment   int
+	effortBatiment int
+}
```
Il faudrait un nouveau fichier pour les batiments dans le dossier `pkg`

gcleroux commented on main.go:
```diff
@@ -2,10 +2,96 @@ package main
+	{idBatiment: 1, nomBatiment: "Parc", prixBatiment: 250, effortBatiment: 20},
+	{idBatiment: 2, nomBatiment: "Hopital", prixBatiment: 500, effortBatiment: 50},
+	{idBatiment: 3, nomBatiment: "Hotel", prixBatiment: 400, effortBatiment: 40}}
+var villeContenu = []string{}
```
C'est plus une note pour le futur, mais il faudrait avoir un fichier YAML pour la configuration. Pour l'instant de le faire comme ca c'est chill imo, mais je vais me mettre la dessus

gcleroux commented on main.go:
```diff
@@ -2,10 +2,96 @@ package main
+		if budget > choixBatiments[batimentChoisi].prixBatiment {
+			commande := choixBatiments[batimentChoisi]
+			budget = budget - commande.prixBatiment
+			OuvrierConstruit(commande)
```
On devrait avoir des workers nodes `ouvriers` qui écoutent sur un channel afin de recevoir les projets de constructions.

En ce moment on a juste 1 ouvrier qui travaille dans la ville

Nost22 replied:
Yup, c'est une autre grosse étape de parallélisme qu'on va pouvoir coder maintenant qu'on a la base du SCAM

gcleroux commented on main.go:
```diff
@@ -2,10 +2,96 @@ package main
+func MayorStart(budget int, nbOuvrier int, choixBatiments []batimentInfo) int {
+	plusPetitPrix := 0
+	go func() {
+		plusPetitPrix = TrouveBatimentMoinsCher()
```
Le call en go-routine ne sert pas a grand chose, on va tester la condition du plus petit prix avant que le retour arrive.

En gros, au premier tour de boucle for, le plus petit prix va etre de 0, mais va s'ajuster quand la go routine va retourner.

Nost22 replied:
"En gros, au premier tour de boucle for, le plus petit prix va être de 0, mais va s'ajuster quand la go routine va retourner."
That's right, c'est l'effet désiré et selon mes tests en débugging, ça fonctionne.

"Le call en go-routine ne sert pas a grand chose, on va tester la condition du plus petit prix avant que le retour arrive."
Je comprends pas ton message. Live, il arrive à la ligne 56 et essaie 1 fois avant que le retour de la fonction TrouveBatimentMoinsCher() change la valeur de "plusPetitPrix". 
Comme ça, il commence la loop pendant que la fonction TrouveBatimentMoinsCher() run, ce qui sauve un peu de temps.

C'est l'effet désiré imo, alors je comprends pas pourquoi tu dis que ça sert pas à grand chose.

gcleroux replied:
> Comme ça, il commence la loop pendant que la fonction TrouveBatimentMoinsCher() run, ce qui sauve un peu de temps.

C'est juste malpropre/potentiellement dangereux de le faire comme ça. Au mieux tu sauves un millionième de seconde, mais tu risque d'avoir des edge cases car la boucle est étroitement couplé avec la valeur du batiment le moins cher

gcleroux reviewed (changes requested):
Good job, ça commence à prendre forme ce projet la!

Il faudrait aussi écrire des tests pour s'assurer que tout fonctionne bien down the line.

cimo2801 reviewed (approved):
Good job !

KevinBissonnette reviewed (approved):
nice !

gcleroux committed:
Reformatting des fichiers en pkg

gcleroux committed:
Generate up-to-date rapport.pdf

gcleroux commented:
@Nost22 J'ai refactoré le code. Faudrait juste faire des tests avant de merge pour s'assurer que tout marche bien.

gcleroux commented on pkg/people/mayor.go:
```diff
@@ -1,5 +1,38 @@
+func MayorStart(budget int, nbOuvrier int) int {
+	plusPetitPrix := batiment.TrouveBatimentMoinsCher()
+
+	for budget > plusPetitPrix {
```
@Nost22 Cette ligne la ne marche pas. Ça devrait être un `>=`

Nost22 replied:
C'est corrigé dans la branche First-Logic, j'assume que tu l'as corrigé dans pool-ouvrier avant que je les merge

gcleroux replied:
Yep

gcleroux committed:
Ouvriers concurrent avec des channels

gcleroux committed:
Generate up-to-date rapport.pdf

Nost22 committed:
Merge pull request #5 from gcleroux/pool-ouvrier

Ouvriers concurrent avec des channels

gcleroux committed:
Merge branch 'main' into First-Logic

gcleroux commented:
J'ai remis la branche à jour avec `main`. 

Il y a un changement dans le CI qui fait qu'il va dire que la job du rapport fail. It's expected, don't worry.
Le seul qui doit absolument réussir c'est les tests.

cimo2801 committed:
commit inital citoyen

cimo2801 committed:
Delete citoyen_test.go

Oubli

cimo2801 committed:
sauvegarde de fin de journée

cimo2801 committed:
Correction bug #7

cimo2801 committed:
Changement nom fonction des citoyens

cimo2801 committed:
Quelques petits correctifs

Nost22 committed:
Merge pull request #8 from gcleroux/citoyen

Citoyen

Nost22 committed:
Ajouté logique Jour, adapté logique ouvrier.

Il manque la logique du Building Board pour bien faire fonctionner les ouvriers.
Écrit les messages et logique d'échec pour les ressources secondaires.

cimo2801 committed:
Defer l'unlock des mutex

Nost22 committed:
Ajouté le Building Board et sa logique dans les Ouvriers.

Bien qu'il reste des bugs, les ouvriers peuvent maintenant construire tous les buildings que le Maire peut commander avec son budget.

Nost22 committed:
Cycle du jour dépend de la fin de tous les threads, pas d'un sleep. BuildingBoard reste dans l'ordre FCFS après un Remove. Ajouté sur la logique des Ouvriers.

Nost22 committed:
Ajout de TauxEmploi et NbEmploye. Nouveau Ouvriers chaque jours (rdm). Logique TauxEmployé

cimo2801 committed:
Merge branch 'JourDeJeu+Ouvrier' into First-Logic
